src/sshagent/SSHAgent.cpp

```diff
@@ -29,46 +29,68 @@
 #include <windows.h>
 #endif
 
-SSHAgent* SSHAgent::m_instance;
+Q_GLOBAL_STATIC(SSHAgent, s_sshAgent);
 
-SSHAgent::SSHAgent(QObject* parent)
-    : QObject(parent)
+SSHAgent::~SSHAgent()
 {
-#ifndef Q_OS_WIN
-    m_socketPath = config()->get("SSHAuthSockOverride", "").toString();
-    if (m_socketPath.isEmpty()) {
-        m_socketPath = QProcessEnvironment::systemEnvironment().value("SSH_AUTH_SOCK");
-    }
-#else
-    m_socketPath = "\\\\.\\pipe\\openssh-ssh-agent";
-#endif
+    removeAllIdentities();
 }
 
-SSHAgent::~SSHAgent()
+SSHAgent* SSHAgent::instance()
 {
-    auto it = m_addedKeys.begin();
-    while (it != m_addedKeys.end()) {
-        // Remove key if requested to remove on lock
-        if (it.value()) {
-            OpenSSHKey key = it.key();
-            removeIdentity(key);
-        }
-        it = m_addedKeys.erase(it);
-    }
+    return s_sshAgent;
 }
 
-SSHAgent* SSHAgent::instance()
+bool SSHAgent::isEnabled() const
 {
-    if (!m_instance) {
-        qFatal("Race condition: instance wanted before it was initialized, this is a bug.");
+    return config()->get("SSHAgent").toBool();
+}
+
+void SSHAgent::setEnabled(bool enabled)
+{
+    if (isEnabled() && !enabled) {
+        removeAllIdentities();
     }
 
-    return m_instance;
+    config()->set("SSHAgent", enabled);
+}
+
+QString SSHAgent::authSockOverride() const
+{
+    return config()->get("SSHAuthSockOverride").toString();
+}
+
+void SSHAgent::setAuthSockOverride(QString& authSockOverride)
+{
+    config()->set("SSHAuthSockOverride", authSockOverride);
+}
+
+#ifdef Q_OS_WIN
+bool SSHAgent::useOpenSSH() const
+{
+    return config()->get("SSHAgentOpenSSH").toBool();
 }
 
-void SSHAgent::init(QObject* parent)
+void SSHAgent::setUseOpenSSH(bool useOpenSSH)
+{
+    config()->set("SSHAgentOpenSSH", useOpenSSH);
+}
+#endif
+
+QString SSHAgent::socketPath(bool allowOverride = true) const
 {
-    m_instance = new SSHAgent(parent);
+    QString socketPath;
+
+#ifndef Q_OS_WIN
+    socketPath = authSockOverride();
+    if (socketPath.isEmpty() || !allowOverride) {
+        socketPath = QProcessEnvironment::systemEnvironment().value("SSH_AUTH_SOCK");
+    }
+#else
+    socketPath = "\\\\.\\pipe\\openssh-ssh-agent";
```

Comment:
This is hard to parse, can you include a comment that describes what is going on here

Reply:
Split it up to make it more apparent what is going on. Added a comment as well.